File: java/02.Java1/06.DEVELOP/TypeDefTest.java
class T {
    public int v;

    public T() {
    }

    public T(int val) {
        v = val;
    }
}

class PT {
    public T v;

    public PT() {
    }

    public PT(T val) {
        v = val;
    }
}

class PPT {
    public PT v;

    public PPT() {
    }

    public PPT(PT val) {
        v = val;
    }
}

class A {
    public static final int SIZE = 10;

    // C의 unsigned char가 char보다는 byte의 의미로 보다 많이 사용된다면,
    // 다음에서 byte[]로 선언하는 것이 더 적절하다.
    public char[] v = new char[SIZE];
}

class PA {
    public A v;

    public PA() {
    }

    public PA(A val) {
        v = val;
    }
}

class TypeDefTest {
    public static final int SIZE = 10;
    public static void main(String[] args) {
        T t1 = new T(), t2 = new T();
        PT pt = new PT(t1);
        PPT ppt = new PPT(pt);
        A a = new A();
        PA pa = new PA(a);

        t1.v = 1;
        t2.v = 2;
        System.out.println("t1.v = " + t1.v + ", t2.v = " + t2.v +
                           ", pt.v.v = " + pt.v.v + 
                           ", ppt.v.v.v = " + ppt.v.v.v );
        pt.v = t2;
        System.out.println("t1.v = " + t1.v + ", t2.v = " + t2.v +
                           ", pt.v.v = " + pt.v.v + 
                           ", ppt.v.v.v = " + ppt.v.v.v );
    
        a.v[0] = 'a';
        System.out.println("a.v[0] = " + a.v[0] +
                           ", pa.v.v[0] = " + pa.v.v[0] );
    }
}

File: java/02.Java1/06.DEVELOP/typedef_test.c
#include <stdio.h>

typedef  int   T;
typedef  T     *PT;
typedef  PT    *PPT;

#define SIZE 10
typedef  unsigned char A[SIZE];
typedef  A  *PA;

main() {
    T  t1, t2;
    PT pt = &t1;
    PPT ppt = &pt;
    A a;
    PA pa = &a;

    t1 = 1;
    t2 = 2;
    printf("t1 = %d, t2 = %d, *pt = %d, **ppt = %d\n",
           t1, t2, *pt, **ppt);
    pt = &t2;
    printf("t1 = %d, t2 = %d, *pt = %d, **ppt = %d\n",
           t1, t2, *pt, **ppt);
    
    a[0] = 'a';
    printf("a[0] = %c, (*pa)[0] = %c\n",
           a[0], (*pa)[0]);
}
